azure/services/identities/client.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package identities
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/Azure/azure-sdk-for-go/services/msi/mgmt/2018-11-30/msi"
+	"github.com/Azure/go-autorest/autorest"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// client wraps go-sdk.
+type client interface {
+	Get(ctx context.Context, resourceGroupName string, resourceName string) (result msi.Identity, err error)
+}
+
+// azureClient contains the Azure go-sdk Client.
+type azureClient struct {
+	userAssignedIdentities msi.UserAssignedIdentitiesClient
+}
+
+var _ client = (*azureClient)(nil)
+
+// newClient creates a new MSI client from subscription ID.
+func newClient(auth azure.Authorizer) *azureClient {
+	c := newUserAssignedIdentitiesClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
+	return &azureClient{c}
+}
+
+// newIdentitiesClient creates a new user-assigned identities client from subscription ID.
+func newUserAssignedIdentitiesClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) msi.UserAssignedIdentitiesClient {
+	userAssignedIdentitiesClient := msi.NewUserAssignedIdentitiesClientWithBaseURI(baseURI, subscriptionID)
+	azure.SetAutoRestClientDefaults(&userAssignedIdentitiesClient.Client, authorizer)
+	return userAssignedIdentitiesClient
+}
+
+// Get the managed service identity.
+func (ac *azureClient) Get(ctx context.Context, resourceGroupName, name string) (msi.Identity, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "identities.azureClient.Get")
+	defer done()
+
+	return ac.userAssignedIdentities.Get(ctx, resourceGroupName, name)
+}
+
+// GetClientID returns the client ID of a managed service identity, given its full URL identifier.
+func GetClientID(ctx context.Context, auth azure.Authorizer, providerID string) string {
```

Comment:
Consider making this func use the `ac *azureClient` receiver and instead of passing in the authorizer and constructing a new client in the func. We can then create a new client from the controller where we need it.